### PR TITLE
Affirm Component

### PR DESCRIPTION
**Pull request by nauaros**
### Changes
Added the Affirm payment method to the DropIn component.
Resolves issue [COIOS-390](https://youtrack.is.adyen.com/issue/COIOS-390)
Added helper method to fill AddressFormView on tests

### CODE REVIEW

File: Adyen.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -1768,30 +1793,31 @@
...
-				E700A72C2549E2F00012B64C /* BLIKPaymentMethod.swift */,
-				E7763A1025F02BD20046371E /* GiftCardPaymentMethod.swift */,
-				F9FE25B32625ECE7001874BB /* CustomDisplayablePaymentMethod.swift */,
+				F9D57529237C6084009C18B5 /* StoredBCMCPaymentMethod.swift */,
 				E71E8F3D257921D10054B03D /* StoredBLIKPaymentMethod.swift */,
-				E788A3B1265565DF00089448 /* EContextPaymentMethod.swift */,
-				5A988BF5265675ED0007F4C0 /* BoletoPaymentMethod.swift */,
+				E224088C22B0FD220058923E /* StoredPayPalPaymentMethod.swift */,
+				F9620D9E23C7457F005209FC /* WeChatPayPaymentMethod.swift */,
 			);
```
Comment (descorp) [inline Adyen.xcodeproj/project.pbxproj:1821]: Did you ordered payment methods alphabetically? :D

File: Adyen/Core/Payment Methods/AffirmPaymentMethod.swift [hunk 2]
```diff
@@ -0,0 +1,40 @@
+//
+//  AffirmPaymentMethod.swift
+//  Adyen
+//
+//  Created by Naufal Aros on 7/6/21.
+//  Copyright © 2021 Adyen. All rights reserved.
+//
```
Comment (descorp) [inline Adyen/Core/Payment Methods/AffirmPaymentMethod.swift:7]: hm.. do you have `swiftformat` ?
This is standard header

Comment (nauaros) [reply Adyen/Core/Payment Methods/AffirmPaymentMethod.swift:7]: I have installed it but it's not been triggered. Gonna take a look at it

File: Tests/AdyenTests/Adyen Tests/Components/Affirm/AffirmComponentTests.swift [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+        let lastNameView: FormTextInputItemView = try XCTUnwrap(view.findView(by: "AdyenComponents.AffirmComponent.lastNameItem"))
+        populate(textItemView: lastNameView, with: "Del Mar")
+        
+        let phoneNumberView: FormPhoneNumberItemView = try XCTUnwrap(view.findView(by: "AdyenComponents.AffirmComponent.phoneNumberItem"))
+        populate(textItemView: phoneNumberView, with: "2025550146")
+        
+        let emailView: FormTextInputItemView = try XCTUnwrap(view.findView(by: "AdyenComponents.AffirmComponent.emailItem"))
+        populate(textItemView: emailView, with: "katrina@mail.com")
+        
+        // ===================== BILLING ADDRESS =====================
```
Comment (nauaros) [inline Tests/AdyenTests/Adyen Tests/Components/Affirm/AffirmComponentTests.swift:173]: I'm aware of this comments. This is WIP

File: Adyen/UI/Form/FormViewController.swift [hunk 4]
```diff
@@ -144,7 +144,16 @@ open class FormViewController: UIViewController, Localizable, KeyboardObserver,
...
     
     private func getAllValidatableItems() -> [ValidatableFormItem] {
-        itemManager.flatItems.compactMap { $0 as? ValidatableFormItem }
+        let unhiddenItems = itemManager.topLevelItem.filter { item in
+            if let hidableItem = (item as? Hidable), hidableItem.isHidden.wrappedValue {
+                return false
+            }
+            
+            return true
+        }
```
Comment (descorp) [inline Adyen/UI/Form/FormViewController.swift:153]: How about 

```swift
let unhiddenItems = itemManager.topLevelItem.filter { ($0 as? Hidable)?.isHidden.wrappedValue == true ? false : true  }
```

File: Adyen.xcodeproj/project.pbxproj (hunk 1 above)
Comment (nauaros) [reply Adyen.xcodeproj/project.pbxproj:1821]: I did. With so many of them it was getting difficult to find Affirm sometimes. Anything wrong?

File: Adyen/UI/Form/Items/Value Pickers/FormRegionPickerItem.swift [hunk 5]
```diff
@@ -37,7 +37,7 @@ internal final class FormRegionPickerItemView: BaseFormPickerItemView<Region> {
     override internal func initialize() {
         let stackView = UIStackView(arrangedSubviews: [titleLabel, inputControl])
         stackView.axis = .vertical
-        stackView.spacing = 8
+        stackView.spacing = 5.5
```
Comment (descorp) [inline Adyen/UI/Form/Items/Value Pickers/FormRegionPickerItem.swift:40]: this is peculiar number 🤔
Is it empirically calculated?

File: AdyenComponents/Affirm/AffirmComponent.swift [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    // MARK: - Items
+    
+    /// :nodoc:
+    private let personalDetailsHeaderItem: FormLabelItem
+    
+    /// :nodoc:
+    public let deliveryAddressToggleItem: FormToggleItem
+    
+    /// :nodoc:
+    public let deliveryAddressItem: FormAddressItem
```
Comment (descorp) [inline AdyenComponents/Affirm/AffirmComponent.swift:22]: Why `deliveryAddressToggleItem` and `deliveryAddressItem ` one is public?

File: AdyenComponents/Affirm/AffirmComponent.swift [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+    
+    // MARK: - Properties
+    
+    private var isDeliveryAddressEnabled: Bool {
```
Comment (descorp) [inline AdyenComponents/Affirm/AffirmComponent.swift:26]: you only use this once, is it really needed? 🤔

File: AdyenComponents/Affirm/AffirmComponent.swift [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+            .firstName,
+            .lastName,
+            .email,
+            .phone,
+            .address,
+            .custom(CustomFormItemInjector(item: FormSpacerItem(numberOfSpaces: 16))),
+            .custom(CustomFormItemInjector(item: deliveryAddressToggleItem)),
+            .custom(CustomFormItemInjector(item: deliveryAddressItem)),
+            .custom(CustomFormItemInjector(item: FormSpacerItem(numberOfSpaces: 24)))
+        ]
```
Comment (descorp) [inline AdyenComponents/Affirm/AffirmComponent.swift:55]: Lines 40-55 could be a private static function?

File: Tests/AdyenTests/Adyen Tests/Components/Affirm/AffirmComponentTests.swift [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+        
+        let submitButton: UIControl = try XCTUnwrap(view.findView(by: "AdyenComponents.AffirmComponent.payButtonItem.button"))
+        submitButton.sendActions(for: .touchUpInside)
+        
+        waitForExpectations(timeout: 10, handler: nil)
+    }
+    
+    // MARK: - Private
+    
+    private func fill(addressView: FormVerticalStackItemView<FormAddressItem>, with address: PostalAddress) {
```
Comment (descorp) [inline Tests/AdyenTests/Adyen Tests/Components/Affirm/AffirmComponentTests.swift:197]: this could be public :)

File: Tests/AdyenTests/Adyen Tests/Components/Affirm/AffirmComponentTests.swift (hunk 3 above)
Comment (descorp) [reply Tests/AdyenTests/Adyen Tests/Components/Affirm/AffirmComponentTests.swift:173]: I guess the comment is fine :D
`+` it is a tests

File: Adyen.xcodeproj/project.pbxproj (hunk 1 above)
Comment (descorp) [reply Adyen.xcodeproj/project.pbxproj:1821]: No! Well done :D
I was just curious why so many changes :)

File: Adyen/Core/Payment Methods/AffirmPaymentMethod.swift (hunk 2 above)
Comment (nauaros) [reply Adyen/Core/Payment Methods/AffirmPaymentMethod.swift:5]: Fixed

File: AdyenComponents/Affirm/AffirmComponent.swift (hunk 6 above)
Comment (nauaros) [reply AdyenComponents/Affirm/AffirmComponent.swift]: Very true. I completely forgot to turn them internal. Thanks

File: Adyen/UI/Form/Items/Value Pickers/FormRegionPickerItem.swift (hunk 5 above)
Comment (nauaros) [reply Adyen/UI/Form/Items/Value Pickers/FormRegionPickerItem.swift:40]: Oh yeah. This is funny cause there's a misalignment between the state picker and the zip code field. The easiest solution is to decrease the spacing to 5.5 so both separator lines have the same Y position.

There's no auto layout involved in positioning both separator lines, so we cannot use that to align them.

I know from a design perspective we work on 8 multiples, but for this case it makes sense to break the rule. Open to suggestions tho

File: AdyenComponents/Affirm/AffirmComponent.swift (hunk 6 above)
Comment (nauaros) [reply AdyenComponents/Affirm/AffirmComponent.swift]: Fixed

File: AdyenComponents/Affirm/AffirmComponent.swift (hunk 8 above)
Comment (nauaros) [reply AdyenComponents/Affirm/AffirmComponent.swift:61]: I thought about it. You mean something like this right?

```
private static func getFields() -> [PersonalInformation] {
   ...
}
```

Comment (descorp) [reply AdyenComponents/Affirm/AffirmComponent.swift:61]: yes, why not :)

File: AdyenComponents/Affirm/AffirmComponent.swift [hunk 10]
```diff
@@ -10,16 +10,22 @@ import UIKit
 /// A component that provides a form for Affirm payment.
 public final class AffirmComponent: AbstractPersonalInformationComponent, Observer {
     
+    private enum ViewIdentifier {
+        static let billingAddress = "billingAddressItem"
+        static let deliveryAddress = "deliveryAddressItem"
+        static let deliveryAddressToggle = "deliveryAddressToggleItem"
+    }
```
Comment (descorp) [inline AdyenComponents/Affirm/AffirmComponent.swift:17]: Nice :D

File: Tests/AdyenTests/Adyen Tests/Components/Affirm/AffirmComponentTests.swift [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+    
+    override func tearDownWithError() throws {
+        paymentMethod = nil
+        apiContext = nil
+        style = nil
+        sut = nil
+        try super.tearDownWithError()
+    }
+    
+    func testInit_shouldPaymentMethodTypeBeAffirm() {
```
Comment (kaphacius) [inline Tests/AdyenTests/Adyen Tests/Components/Affirm/AffirmComponentTests.swift:36]: can we remove `_` from the function name?

File: Adyen/UI/Form/Items/Value Pickers/FormRegionPickerItem.swift (hunk 5 above)
Comment (descorp) [reply Adyen/UI/Form/Items/Value Pickers/FormRegionPickerItem.swift:40]: If it look good - it is on :)

File: Adyen/Core/Payment Methods/AffirmPaymentMethod.swift [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+// This file is open source and available under the MIT license. See the LICENSE file for more info.
+//
+
+import Foundation
+
+/// Affirm payment method.
+public struct AffirmPaymentMethod: PaymentMethod {
+    
+    /// :nodoc:
+    public var type: String
```
[the remark this answers is not in the document]
Comment (nauaros) [reply Adyen/Core/Payment Methods/AffirmPaymentMethod.swift]: Fixed

File: AdyenComponents/Affirm/AffirmComponent.swift (hunk 7 above)
Comment (nauaros) [reply AdyenComponents/Affirm/AffirmComponent.swift:32]: Removed to use toggle value direcly

File: AdyenComponents/Affirm/AffirmComponent.swift (hunk 8 above)
Comment (nauaros) [reply AdyenComponents/Affirm/AffirmComponent.swift:55]: I thought about it but I'm afraid it's not possible since I need to have access to the first name, last name, email and phone fields from the parent class (`AbstractPersonalInformationComponent`). Cannot access them from a static method.

File: AdyenComponents/Affirm/AffirmComponent.swift (hunk 10 above)
Comment (mohammedDehairy) [reply AdyenComponents/Affirm/AffirmComponent.swift:17]: Yeah, Nice!

**Review by mohammedDehairy — COMMENTED** [review]
Really Nice! Naufal

File: Tests/AdyenTests/Adyen Tests/Components/Affirm/AffirmComponentTests.swift (hunk 9 above)
Comment (nauaros) [reply Tests/AdyenTests/Adyen Tests/Components/Affirm/AffirmComponentTests.swift:184]: Yeah, gonna move this to an `XCTestCase` extension so it's available across tests

File: AdyenComponents/Affirm/AffirmComponent.swift [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+        deliveryAddressToggleItem = FormToggleItem(style: style.toggle)
+        deliveryAddressItem = FormAddressItem(initialCountry: Locale.current.regionCode ?? "US", style: style.addressStyle)
+        
+        let fields: [PersonalInformation] = [
+            .custom(CustomFormItemInjector(item: personalDetailsHeaderItem.addingDefaultMargins())),
+            .firstName,
+            .lastName,
+            .email,
+            .phone,
+            .address,
```
Comment (mohammedDehairy) [inline AdyenComponents/Affirm/AffirmComponent.swift:50]: I am wondering if you can use two `.address(title)` items instead of making a `CustomFormItemInjector ` with another `FormAddressItem `, but this can be in a different PR

**Review by mohammedDehairy — APPROVED** [review]

File: AdyenComponents/Affirm/AffirmComponent.swift (hunk 13 above)
Comment (nauaros) [reply AdyenComponents/Affirm/AffirmComponent.swift:50]: Actually it's a good idea. We can update the `addressItemInjector` on the `AbstractPersonalInformationComponent` so we pass the title

**Review by descorp — DISMISSED** [review]

**Review by descorp — APPROVED** [review]

**Review by mohammedDehairy — APPROVED** [review]
